flink-spark-bundle/src/test/java/org/apache/iceberg/flinkspark/TestFlinkSpark.java

```diff
@@ -0,0 +1,348 @@
+
+        builder.equalityFieldColumns(Lists.newArrayList("type"));
+        assertThat(builder.checkAndGetEqualityFieldIds())
+                .containsExactlyInAnyOrder(table.schema().findField("type").fieldId());
+    }
+
+    @TestTemplate
+    public void testEqualityDeleteWritesOnSpark() throws Exception {
+        // Step 1: Write initial data using Flink
+        DataStream<Row> dataStream = env.fromCollection(
+                ImmutableList.of(
+                        row("+I", 1, "value1"),
+                        row("+I", 2, "value2"),
+                        row("+I", 3, "value3")
+                )
+        );
+
+
+        FlinkSink.forRow(dataStream, SimpleDataUtil.FLINK_SCHEMA)
+                .tableLoader(tableLoader)
+                .writeParallelism(parallelism)
+                .append();
+
+        // Execute the Flink job to write initial data
+        env.execute("Write Initial Data");
+
+        // Step 2: Apply equality deletes using Flink
+
+        DataStream<Row> deleteStream = env.fromCollection(
+                ImmutableList.of(
+                        row("-D", 1, "value1"),  // Equality delete row with id=1
+                        row("-D", 2, "value2")   // Equality delete row with id=2
+                )
+        );
+
+        FlinkSink.forRow(deleteStream, SimpleDataUtil.FLINK_SCHEMA)
+                .tableLoader(tableLoader)
+                .equalityFieldColumns(Lists.newArrayList("id", "data"))
+                .writeParallelism(parallelism)
+                .upsert(true)  // Enable UPSERT mode for equality deletes
+                .append();
+
+        // Execute the Flink job to apply equality deletes
+        env.execute("Apply Equality Deletes");
+
+        // Step 3: Use Spark to read the table and verify that equality deletes were applied correctly
+
+        // Initialize SparkSession
+        SparkSession spark = SparkSession.builder()
+                .appName("IcebergSparkRead")
+                .master("local[*]")
+                .config("spark.sql.catalog.hadoop_catalog", "org.apache.iceberg.spark.SparkCatalog")
+                .config("spark.sql.catalog.hadoop_catalog.type", "hadoop")
+                .config("spark.sql.catalog.myCatalog.warehouse", "file:///path/to/warehouse")
+                .getOrCreate();
+
+        // Read the table using Spark
+        Dataset<org.apache.spark.sql.Row> result = spark.read()
+                .format("iceberg")
+                .load("hadoop_catalog." + DATABASE + "." + TestFixtures.TABLE);
+
+        // Collect the result
+        List<org.apache.spark.sql.Row> actualData = result.collectAsList();
+
+        // Expected result after applying equality deletes (only id=3 should remain)
+        List<org.apache.spark.sql.Row> expectedData = ImmutableList.of(RowFactory.create(3, "value3"));
+
+        // Assert that only row with id=3 remains in the table
+        assertThat(actualData).containsExactlyInAnyOrderElementsOf(expectedData);
+
```

Comment:
This is a good test for writing the equality delete using the stream execution environment! I'd also suggest creating a test for an `UPSERT` case by leveraging the `FlinkTableEnviroment`. This will help you see how, the UPSERT leverages an equality delete to replace the value.

You can add a test to [TestFlinkCatalogTable.java](https://github.com/codebase-berkeley/iceberg/blob/main/flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogTable.java) like this:


```
    sql(
        "CREATE TABLE test_table (id INT, data STRING, PRIMARY KEY(id) NOT ENFORCED) WITH ('format-version'='2', 'write.upsert.enabled'='true')");
  

    sql("INSERT INTO test_table VALUES (1, 'a'), (2, 'b'), (3, 'c')");

    // Perform upsert operation
    sql("INSERT INTO test_table VALUES (2, 'updated_b'), (4, 'd')");

```

Reply:
To expand upon this, I'd suggest adding some DeleteFile assertions. For example, what do you expect the DeleteFile to look like we want to delete based on:
* one column:  id = 1
* all columns: id = 1 and data = 'a'
* range: id > 3